### ISSUE TITLE
Does not detect redundant use of collect + iter + map

### ISSUE THREAD

**Issue by EdmundsEcho** [rust-lang/rust-clippy#4154]
There might be reasons this is not always the case, however, it seems like a situation when I was being far too verbose and an opportunity for clippy to let me know.  Here is the verbose version of the code:
```
   (1..=64)
        .collect::<Vec<_>>()
        .iter()
        .map(|&sq| square(sq))
        .sum()

```

Here is the pithy version of the code:
```
    (1..=64).map(square).sum()

```
\- E

**Comment by oli-obk** [comment]
This may be impossible to do if the initial iterator borrows something that the `map` closure wants to use:

```rust
let mut x = [5, 6, 7, 0, 0, 0, 0, 0, 0, 0];
let iter = x.iter().copied();
let n: u32 = iter.collect::<Vec<usize>>()
    .iter()
    .map(|&y| { x[y] = 20; 1 })
    .sum();
println!("{:?}, {:?}", n, x);
```

we have to tread carefully here
